File: alg/strings/two-two.swift
/// Two Two
/// TODO: link

import Foundation

public func twoTwo() {

    // Compute and store strings for powers of 2
    let maxPower: Int = 800
    let powers = computePowers(maxPower)
    let powerStrings = powers.map { p in
        return String(format: "%.0f", arguments: [p])
    }

    // create trie from prefixes
    let trie: Trie<String> = Trie.buildTrie(powerStrings)

    // loop through test cases
    let n: Int = readLn()
    var ans: [Int] = []
    for _ in 0..<n {
        let string: String = getLine()
        let count = trie.countMatches(string)
        println(count)
    }
}

private func computePowers(maxPower: Int) -> [Double] {
    var powers: [Double] = []
    for p in 0...maxPower {
        powers.append(pow(2, Double(p)))
    }
    return powers
}

final class Trie <T: CollectionType where T.Generator.Element: Hashable, T.Generator.Element: Equatable> {
    typealias Value = T.Generator.Element
    typealias Generator = T.Generator
    typealias Output = T

    let value: Value?   // Root node has no value
    var label: Output? = nil
    var output: [Output] = []
    private var rootGoto: Trie? = nil
    private var fail: Trie? = nil
    private var nodes: [Value: Trie] = [:]

    init(value: Value? = nil) {
        self.value = value
    }

    class func buildTrie(superSet: [T]) -> Trie {
        let rootNode = Trie()
        // Root node goto to self when no matching node
        rootNode.rootGoto = rootNode

        // Phase I: build initial trie
        for set in superSet {
            var node = rootNode
            for element in set {
                node = node.addNode(element)
            }
            // Last node needs to have label and output set
            node.output = [set]
            node.label = set
        }

        // Phase II: construct goto and fail functions
        var queue: [Trie] = []
        for (key, nextNode) in rootNode.nodes {
            nextNode.fail = rootNode
            queue.append(nextNode)
        }

        while !queue.isEmpty {
            let node = queue.removeAtIndex(0)
            for (element, nextNode) in node.nodes {
                queue.append(nextNode)

                var failNode = node.fail!
                while failNode.goto(element) == nil {
                    failNode = failNode.fail!
                }
                nextNode.fail = failNode.goto(element)
                nextNode.output += nextNode.fail!.output
            }
        }

        return rootNode
    }

    func addNode(value: Value) -> Trie {
        if let nextNode = nodes[value] {
            return nextNode
        } else {
            let node = Trie(value: value)
            nodes[value] = node
            return node
        }
    }

    private func goto(value: Value) -> Trie? {
        return nodes[value] ?? rootGoto
    }

    func countMatches(collection: T) -> Int {
        var count = 0
        var node = self

        for element in collection {
            while node.goto(element) == nil {
                node = node.fail!
            }
            node = node.goto(element)!
            count += node.output.count
        }

        return count
    }
}

#if CLI_BUILD
    import StdIO
    // silence the xcode, it doesn't like just calling funcName() directly
    let x: Void = twoTwo()
#endif

File: alg/strings/two-two.cpp
/// Two Two
/// TODO: link

#include <iostream>
#include <map>
#include <vector>
#include <string>
#include <queue>
#include <cassert>
#include <cmath>
using namespace std;

#if !defined(CLI_BUILD)
#include "HackerRank.h"
#endif

typedef vector<string> PrefixSet;
typedef PrefixSet::iterator PrefixSetIterator;


class Trie {
public:
    typedef char ValueType;
    typedef map<ValueType, Trie*> TrieMap;
    typedef TrieMap::iterator TrieMapIterator;

    bool root;
    ValueType value;
    string label;
    int count;
    Trie *failNode;
    TrieMap nodes;

    Trie(int value = 0, bool root = false) {
        this->value = value;
        this->root = root;
        this->count = 0;
        this->failNode = NULL;
    }

    Trie *valueNode(ValueType value) {
        TrieMapIterator nodeIt = nodes.find(value);
        return (nodeIt != nodes.end() ? nodeIt->second : NULL);
    }

    Trie *addNode(ValueType value) {
        Trie *node = valueNode(value);
        if (node) {
            return node;
        } else {
            Trie *trie = new Trie(value);
            nodes[value] = trie;
            return trie;
        }
    }

    Trie *gotoNode(ValueType value) {
        Trie *node = valueNode(value);
        // Root node always goto to self if no valid path available
        return node ? node : (root ? this : NULL);
    }

    static Trie &buildTrie(PrefixSet &prefixSet) {
        Trie *rootNode = new Trie(0, true);

        // Phase I: build initial trie
        for (PrefixSetIterator it = prefixSet.begin(); it != prefixSet.end(); ++it) {
            Trie *node = rootNode;
            string prefix = *it;
            for (int i = 0; i < prefix.length(); i++) {
                node = node->addNode(prefix[i]);
            }
            node->label = prefix;
            node->count = 1;
        }

        // Phase II: construct goto and fail functions
        queue<Trie*> nodeQueue;
        for (TrieMapIterator it = rootNode->nodes.begin(); it != rootNode->nodes.end(); ++it) {
            Trie *nextNode = it->second;
            nextNode->failNode = rootNode;
            nodeQueue.push(nextNode);
        }

        while (!nodeQueue.empty()) {
            Trie *node = nodeQueue.front();
            nodeQueue.pop();

            for (TrieMapIterator it = node->nodes.begin(); it != node->nodes.end(); ++it) {
                ValueType value = it->first;
                Trie *nextNode = it->second;
                nodeQueue.push(nextNode);

                Trie *failNode = node->failNode;
                while (failNode->gotoNode(value) == NULL) {
                    failNode = failNode->failNode;
                }

                nextNode->failNode = failNode->gotoNode(value);
                nextNode->count += nextNode->failNode->count;
            }
        }

        return *rootNode;
    }

    int countMatches(string str) {
        int count = 0;
        Trie *node = this;

        for (int i = 0; i < str.length(); i++) {
            ValueType value = str[i];
            while (!node->gotoNode(value)) {
                node = node->failNode;
            }
            node = node->gotoNode(value);
            count += node->count;
        }

        return count;
    }
};

PrefixSet &buildPrefixSet(int maxPower) {
    PrefixSet *set = new PrefixSet();
    for (int p = 0; p <= maxPower; p++) {
        double po2 = pow(2, p);
        string po2String = to_string(po2);
        size_t lastDot = po2String.find_last_of(".");
        if (lastDot != string::npos) {
            po2String = po2String.substr(0, lastDot);
        }
        set->push_back(po2String);
    }

    return *set;
}

void twoTwoCpp() {
    PrefixSet prefixSet = buildPrefixSet(800);

    Trie trie = Trie::buildTrie(prefixSet);

    int n;
    cin >> n;
    for (int i = 0; i < n; i++) {
        string input;
        cin >> input;
        int count = trie.countMatches(input);
        cout << count << endl;
    }
}

#ifdef CLI_BUILD
int main() {
    twoTwoCpp();
    return 0;
}
#endif